### ISSUE TITLE
Checkbox table

### ISSUE THREAD

**Issue by dt-mafe** [dynatrace-oss/barista#1052]
# Feature Request

## Summary

Create a column that allows you to select rows in a table similar to the already existing favorite column. The column should show a checkbox for each row and in the column header. The table data source should keep track of selection changes.

## Feature Description

The new column should show check-boxes for each row. Optionally, it should also show a checkbox in the table header which allows selecting/deselecting all rows at once.
The column has to support the following scenarios:

- Row and header checkboxes should support the selected/unselected and indeterminate state
- The column header should be able to contain a default header label, checkbox or both
- You should be able to mark certain rows as not selectable/disabled
- The column should emit an event when a checkbox is clicked

To make using the column easier and avoid duplicate selection handling implementations, the DtTableDataSource should get support for handling selection changes. It should also be possible to use the checkbox column without using the default selection handling and using your own custom implementation instead. Also the selection handling in the data-source should not be specific to the checkbox column and also be usable with other columns providing similar functionality (e.g. the favorite column).
Therefore I would suggest using a directive to enable the default selection handling in the data source that couples a selectable column with the data source and triggers selection updates in the data source.

The selection handling should handle the following scenarios:
- Update selection when checkbox of a row clicked
- Select/Deselect all selectable rows when the header checkbox is clicked
- Update the checkbox column header based on current selection
- Getting the current selection
- Updating the current selection
- Limiting how many rows can be selected
- Limiting which rows can be selected

## Attachments

<img width="637" alt="location_table_gwt" src="https://user-images.githubusercontent.com/23524331/82312711-b99f7c00-99c7-11ea-810f-a39d8c81ecc3.png">
<img  src="https://user-images.githubusercontent.com/23524331/82341057-3cd4c800-99f0-11ea-93ea-a853e62f26fe.png">

**Comment by lukasholzer** [comment]
Hey, thank you very much for filing this issue! Are you willing to contribute this to the library?
If you want to please use this issue and provide an API proposal for the component.

As a reference this would be a good API proposal: https://github.com/dynatrace-oss/barista/issues/673#issuecomment-612874961

**Comment by ffriedl89** [comment]
@dt-mafe Thanks for creating this. I think you outlined the plan pretty well already. I think the only thing missing is the components, inputs, outputs and method listed

**Comment by dt-mafe** [comment]
### API Proposal

#### Checkbox column
```
<dt-checkbox-column [showHeaderCheckbox]="true"  [allSelected]="false" [anySelected]="true" (rowCheckboxClicked)="fn($event)"  (headerCheckboxClicked)="fn($event) (selectionToggled)="fn2($event)"></dt-checkbox-column>
```

#### Inputs

Name | Type | Default value | Description
-- | -- | -- | --
showHeaderCheckbox| boolean | true | Whether a checkbox is shown in the column header or not
allSelected| boolean | false | Determines the header checkbox checked value
anySelected| boolean | false | Determines the header checkbox indeterminate value

#### Outputs

Name | Type | Description
-- | -- | --
rowCheckboxClicked| EventEmitter<DtCheckboxChange<T>> | Event emitted when the checkbox of a table row is clicked (not sure if this is really needed)
headerCheckboxClicked| EventEmitter<DtCheckboxChange<T>> | Event emitted when the header checkbox is clicked (not sure if this is really needed)
selectionToggled  | EventEmitter<T \| null> | Event emitted when row or header selection changes, either emits the value of the row that was clicked or null if the header checkbox was clicked


#### Row data
In order to determine the state of a checkbox in a table row the checkbox column supports either a displayAccessor or the required object being set on the table row value under a property using the same as the column name.
The object provided by the displayAccessor or on the row-data has three boolean flags to determine the checkbox state:

Name | Description
-- | --
disabled | Whether the checkbox should be shown as disabled or not
checked | Whether the checkbox is checked or not
indeterminate | Whether the checkbox is in the indeterminate state or not

#### DtSelection directive
selector: dtSelection
Has to be added to a selectable column and will then subscribe to events of that column and in turn emit events based on events from the column.

#### Inputs

Name | Type | Default value | Description
-- | -- | -- | --
selectionLimit | number\|undefined | undefined | Limits how many records can be selected at once, unlimited if undefined
selectable | Predicate<T> | () => true | Predicate to determine if a row can be selected or not

#### Outputs
Name | Type | Description
-- | -- | --
selectionChange | EventEmitter<T \| null> | Emits event when a checkbox in the column is clicked with the value of the row that was clicked or null if the header checkbox was clicked

#### Table data source

The table data source gets some new properties:

Name | Type | Description
-- | -- | --
selectionModel | SelectionModel<T> | Stores current selection and provides methods to change that selection
selection | DtSelection<T> | The directive forwarding events from the column to the data source

Example usage:

```
example.html:

<dt-table [dataSource]="dataSource" dtSort>
  <dt-checkbox-column
    name="checkbox"
    [showHeaderCheckbox]="true"
    dtSelection
    [selectionLimit]="5"
    [selectable]="isSelectable"
  ></dt-checkbox-column>
  <dt-simple-text-column name="host" label="Host"></dt-simple-text-column>
  <dt-simple-number-column name="cpu" label="CPU"></dt-simple-number-column>
  <dt-header-row *dtHeaderRowDef="['checkbox', 'host', 'cpu']"></dt-header-row>
  <dt-row *dtRowDef="let row; columns: ['checkbox', 'host', 'cpu']"></dt-row>
</dt-table>

Current selection: {{getCurrentSelection()}}

example.ts 
...
  @ViewChild(DtSelection, { read: DtSelection, static: true })
  selection: DtSelection<Row>;

  // Initialize the table's data source
  dataSource: DtTableDataSource<Row>;
  constructor() {
    this.dataSource = new DtTableDataSource(this.data);
  }

  ngAfterViewInit(): void {
    // Set the dtSelection reference on the data source to enable default selection handling
    this.dataSource.selection = this.selection;
  }

  getCurrentSelection(): string {
    return this.dataSource.selectionModel.selected.map((data: Row) => data.host).join(", ");
  }

  isSelectable(entry: { host: string; cpu: number }): boolean {
    return !entry.host.endsWith('(disabled)');
  }
...
```
Result:
![image](https://user-images.githubusercontent.com/23524331/82451986-00b46c80-9aaf-11ea-9cc2-b7c22e29cd60.png)

#### Additional notes:
The approach outlined above only supports one selectable column per table for the default selection handling. I am not sure if we want to support multiple columns there, but if we do then the data source would contain a Map<string, SelectionModel<T>> instead of just the single selection model where the column name is used as key for the map.

**Comment by ffriedl89** [comment]
After a lenghty investigation and issues that came up during development of the checkbox-column and changedetection issues we decided to go a bit of a different route. 

The initial idea was to have a checkbox-column similar to our other simple-column implementations. The CdkTable handles CD very strangely for those nested columns that contain the cellDefs internally. This caused ExpressionChangedAfterChecked errors and made the implementation basically feel hacky. @dt-mafe found a workaround but the solution never felt right. 

After more investigations we came up with a solution that solves the problem of lengthy repetitions in the template and reusability. 

The new API would look like
```
<dt-table [dataSource]="dataSource" dtSelection>
    <ng-container dtColumnDef="select">
      <dt-table-header-checkbox *dtHeaderCellDef></dt-table-header-checkbox>
      <dt-table-cell-checkbox *dtCellDef="let row" [row]="row"></dt-table-cell-checkbox>
    </ng-container>
...
</dt-table>
```
The DtSelection directive was moved to the table and it handles the selectionModel used. The dt-table-header-checkbox and the dt-table-cell-checkbox (names to be discussed) handle the checkbox logic of single and multi selection. 

This does not cause any CD issues and reduces the code necessary quite significantly. 

Thanks @dt-mafe for the patience, help and time invested!
